JS: recognize .split("?")[0] in more cases for URL redirection

erik-krogh opened the pull request:
Fixes [these FPs](https://lgtm.com/projects/g/golden-layout/golden-layout/alerts?id=js%2Fclient-side-unvalidated-url-redirection&mode=list) that showed up in an anomalous query results a while ago. 

The issue is we didn't recognize the `.split("?")[0]` pattern if there was any data-flow. E.g. 

```JavaScript
var parts = document.location.href.split( '?' );
var safe = parts[0];
```

With this PR a `.getALocalSource()` is added, such that we recognize the safe way of constructing a redirect URL in more cases. 

The pattern is not common, and I found no instances of it on our [standard 236 benchmarks](https://lgtm.com/query/2241866304612215393/).   

I know it looks like an ad-hoc addon, but the alternative solution would be to add another flow-label, and I don't think that would perform well.

esbena commented:
> the alternative solution would be to add another flow-label, and I don't think that would perform well.

I think the `split-and-destruct` pattern is common enough that you can generalize `QueryPrefixSanitizer` to be a class we may be able to use elsewhere (`ag '"split"' -G qll` gives some results), that is another alternative that avoids the adhoc feeling.

I am thingking of something like:
```ql
class StringSplitCall extends DataFlow::MethodCallNode {
  string getSplitAt() { getArgument(0).mayHaveStringValue(result) }
  DataFlow::Node getUnsplit() { result = getReceiver() }
  DataFlow::Node getAnElementRead(int i) { ... }
}
```

asgerf commented:
I believe this also affects the DOM-based XSS query. Might be worth sharing the sanitizer if doing so doesn't have any adverse side effects.

esbena commented on change-notes/1.25/analysis-javascript.md:
```diff
@@ -20,6 +20,7 @@
 | Misspelled variable name (`js/misspelled-variable-name`) | Message changed | The message for this query now correctly identifies the misspelled variable in additional cases. |
 | Uncontrolled data used in path expression (`js/path-injection`) | More results | This query now recognizes additional file system calls. |
 | Uncontrolled command line (`js/command-line-injection`) | More results | This query now recognizes additional command execution calls. |
+| Client-side URL redirect (`js/client-side-unvalidated-url-redirection`) | Less results | This query now recognizes additional safe patterns of doing URL redirects. |
```
This  should be expanded to mention the other affected queries.

esbena commented on javascript/ql/src/semmle/javascript/StandardLibrary.qll:
```diff
@@ -146,3 +146,37 @@ class StringReplaceCall extends DataFlow::MethodCallNode {
+class StringSplitCall extends DataFlow::MethodCallNode {
+  StringSplitCall() {
+    this.getMethodName() = "split" and
+    (getNumArgument() = 1 or getReceiver().mayHaveStringValue(_))
```
The signature is: `str.split([separator[, limit]])`
(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split)

esbena commented on javascript/ql/src/semmle/javascript/StandardLibrary.qll:
```diff
@@ -146,3 +146,37 @@ class StringReplaceCall extends DataFlow::MethodCallNode {
+  /**
+   * Gets a the SourceNode for the string before it is split.
+   */
+  DataFlow::SourceNode getUnsplit() { result = getReceiver().getALocalSource() }
```
I dislike the `getUnsplit` name (even though it is my own suggestion). 

Perhaps `getStringSource`, and dually `getASubstringRead(int i)` below would be better?

This has no result for `"foo|bar".split("|").map(...)`. I can live with that if a renaming like the above is implemented.
If we need to reason about the shorthand above, we can introduce `string getString()` and `string getSubstring(int i)`.

asgerf replied:
In the `StringOps` module I tried to stick with `getBaseString` as it was meaningful yet generic enough to be used consistently across the different string operations. I think we can use the same here, and possibly drop the `getALocalSource`.

erik-krogh replied:
I went with `getBaseString`, and dropped the `getALocalSource()`.

asgerf commented on javascript/ql/src/semmle/javascript/StandardLibrary.qll:
```diff
@@ -146,3 +146,37 @@ class StringReplaceCall extends DataFlow::MethodCallNode {
+  /**
+   * Gets a string that determines where the string is split.
+   */
+  string getSplitAt() {
```
I didn't want to block your PR on naming issues like this, but since we're at it:
```suggestion
  string getDelimiter() {
```

esbena replied:
```suggestion
  string getSeparator() {
```
(matches the MDN signature above)

asgerf replied:
Either is fine with me

esbena commented on javascript/ql/src/Security/CWE-400/PrototypePollutionUtility.ql:
```diff
@@ -23,8 +23,8 @@ import semmle.javascript.DynamicPropertyAccess
-    getSplitAt() = "." and
-    getUnsplit() instanceof ParameterNode
+    getSeparator() = "." and
+    getBaseString() instanceof ParameterNode
```
this needs a `flowsTo`, doesn't it? If so, please add a test that would have caught it.

erik-krogh replied:
An existing test did catch it, and I fixed it while you were writing that comment.

esbena commented on change-notes/1.25/analysis-javascript.md:
```diff
@@ -20,6 +20,10 @@
 | Uncontrolled data used in path expression (`js/path-injection`) | More results | This query now recognizes additional file system calls. |
 | Uncontrolled command line (`js/command-line-injection`) | More results | This query now recognizes additional command execution calls. |
+| Client-side URL redirect (`js/client-side-unvalidated-url-redirection`) | Less results | This query now recognizes additional safe patterns of doing URL redirects. |
+| Client-side cross-site scripting (`js/xss`) | Less results | This query now recognizes more safe strings based on URLs. |
+| Incomplete URL scheme check (`js/incomplete-url-scheme-check`) | More results | This query now recognizes more url scheme checks. |
+| Prototype pollution in utility function (`js/prototype-pollution-utility`) | More results | This query now recognizes more utility functions vulnerable to prototype polution. |
```
```suggestion
| Client-side cross-site scripting (`js/xss`) | Less results | This query now recognizes additional safe strings based on URLs. |
| Incomplete URL scheme check (`js/incomplete-url-scheme-check`) | More results | This query now recognizes additional url scheme checks. |
| Prototype pollution in utility function (`js/prototype-pollution-utility`) | More results | This query now recognizes additional utility functions as vulnerable to prototype polution. |
```

erik-krogh commented:
An [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1588789669900) came back with not great performance, and an FP for the following pattern: 

```JavaScript
window.location.replace(window.location.href.split("#")[0] + "#mappage");
```

I fixed the FP, and I'll look into the performance.

erik-krogh commented:
A [security/nightly](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1589610589416) evaluation came back with mixed results.  
But a [redo of the 10 worst looks OK](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1589632486644). 

So I think this is ready to land.
